core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -149,6 +155,67 @@ class FileMessageSet private[kafka](@volatile var file: File,
   }
 
   /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   *
+   * The search will stop immediately when it sees a message in format version before 0.10.0. This is to avoid
+   * scanning the entire log when all the messages are still in old format.
+   *
+   * @param targetTimestamp The timestamp to search for.
+   * @param startingPosition The starting position to search.
+   * @return None, if no message exists at or after the starting position.
+   *         Some(TimestampOffset) with (largest_timestamp_checked, the_next_offset_to_read) otherwise.
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] = {
+    var maxTimestampChecked = Message.NoTimestamp
+    var lastOffsetChecked = -1L
+    val messagesToSearch = read(startingPosition, sizeInBytes)
+    for (messageAndOffset <- messagesToSearch) {
+      val message = messageAndOffset.message
+      lastOffsetChecked = messageAndOffset.offset
+      // Stop searching once we see message format before 0.10.0.
+      // This equivalent as treating message without timestamp has the largest timestamp.
+      // We do this to avoid scanning the entire log if no message has a timestamp.
+      if (message.magic == Message.MagicValue_V0)
+        return Some(TimestampOffset(maxTimestampChecked, messageAndOffset.offset))
```

Comment:
If we return 105 when search for 1500, would that cause problem for the transitional period? Previously user will consume from 100, which may be the messages they actually wanted. Once they switch to new timestamp search, the user will not see message 100 - 104 any more. In our example there are only 5 messages, but in real world there could be more.

It looks there is a trade-off we need to make between backward compatibility and accuracy of search. I am not sure how critical the backward compatibility here is. Personally I feel that when there are mixture of messages with and w/o timestamps, backward compatibility seems more important. Because the accuracy will increase when more and more messages contains timestamp. And we will have the full accuracy once the users have fully rolled out the new message format.